### PR TITLE
Implement `fingerprint` method in Core classes

### PR DESCRIPTION
**Pull request by ancalita**
**Proposed changes**:
- Implements #9133 
- I implemented the fingerprint method for `Event`, `Slot` and `TrackerWithCachedStates`.
- This already existed in `Domain` and `StoryGraph` so left untouched.
- Instead added some missing tests for `Domain.fingerprint()` - `StoryGraph` had an existing test in `test_graph.py`

**Status (please check what you already did)**:
- [X] added some tests for the functionality
- [ ] updated the documentation
- [x] updated the changelog (please check [changelog](https://github.com/RasaHQ/rasa/tree/main/changelog) for instructions)
- [X] reformat files using `black` (please check [Readme](https://github.com/RasaHQ/rasa#code-style) for instructions)

### CODE REVIEW

File: tests/shared/core/test_generator.py
Comment (ancalita) [inline tests/shared/core/test_generator.py:73]: Not very pleased with accessing a private attribute here, I didn't know how to update the cached states otherwise🤔

File: tests/shared/core/test_events.py
Comment (ancalita) [inline tests/shared/core/test_events.py:741]: Here I wasn't sure if to use `rasa.shared.utils.common.all_subclassess` to iterate through all the subclassess of Event in case we add Event types in the future? This would imply swapping the long list in `pytest.mark.parametrize` with a long if-elif because some events have some specific arguments that need to be passed.

File: rasa/shared/core/generator.py
Comment (joejuzl) [inline rasa/shared/core/generator.py:231]: @twerkmeister added functionality for `.fingerprint` to `deep_container_fingerprint` in https://github.com/RasaHQ/rasa/pull/9175.

File: rasa/shared/core/generator.py
Comment (joejuzl) [inline rasa/shared/core/generator.py:235]: Maybe `data` should be a `dict` rather than a list?. We probably shouldn't rely on `slots.values()` always giving the same order, and we don't want that to change the fingerprint.

File: rasa/shared/core/generator.py
Comment (joejuzl) [inline rasa/shared/core/generator.py:220]: Can any of this functionality be extracted into the parent class `DialogueStateTracker`?

File: rasa/shared/core/generator.py
Comment (joejuzl) [inline rasa/shared/core/generator.py:237]: I'm not an expert on this area of the code, but I wonder if we need to fingerprint this stuff if we just fingerprinted `self.events` instead (maybe in the base class)?

File: rasa/shared/core/slots.py
Comment (joejuzl) [inline rasa/shared/core/slots.py:158]: I'm not sure this is enough. What about things like `influence_conversation`? and for subclasses like `CategoricalSlot` we have things like `self.values`.

File: tests/shared/core/test_domain.py
Comment (joejuzl) [inline tests/shared/core/test_domain.py:1482]: What about slots, entities, forms etc.?

File: tests/shared/core/test_generator.py
Comment (joejuzl) [reply tests/shared/core/test_generator.py:73]: Does the cached states do anything other than just speed up certain methods? I.e. does it actually hold any unique information not encapsulated by the events and slots etc.?

**Review by joejuzl — CHANGES_REQUESTED** [review]

File: rasa/shared/core/generator.py
Comment (ancalita) [reply rasa/shared/core/generator.py:220]: Seems that `sender_id` and `slots` are common attributes - should I create a fingerprint of `DialogueStateTracker` using these attributes? Then how do I combine the resulting hash of `super().fingerprint()` in `TrackerWithCachedStates` with the its other attributes (`domain`, `_states_for_hashing`)?

File: rasa/shared/core/generator.py
Comment (ancalita) [reply rasa/shared/core/generator.py:227]: Ah true, no need to hash it then and there - I can just append it to `data` directly.

File: rasa/shared/core/generator.py
Comment (ancalita) [reply rasa/shared/core/generator.py:231]: I can see the PR hasn't been merged it so I can't use this added functionality at the moment. 
Would it be fine to actually change this to:
```
data.append(self.domain.as_dict())
```

Comment (joejuzl) [reply rasa/shared/core/generator.py:231]: it's only the 4 lines in this file: https://github.com/RasaHQ/rasa/pull/9175/files#diff-d0c59274cb6aad666a5addf9f758510ba9a2ecdcb48b08a16b904f90c980e017R221

So probably easier to just copy the changes across to this branch.

File: tests/shared/core/test_generator.py
Comment (ancalita) [reply tests/shared/core/test_generator.py:73]: At a first look [here](https://github.com/RasaHQ/rasa/blob/2039cef57780f034da22efcf8c1534d12f3d0ac3/rasa/shared/core/domain.py#L1078) - seems that most of the information here should be captured in events, domain, slots. I can remove this and instead add `self.events` to fingerprintable data?

File: tests/shared/core/test_events.py
Comment (ancalita) [reply tests/shared/core/test_events.py:741]: Yup, can do - however a warning that not all subclassess exist in this list list, for example internal events such as `WronglyPredictedAction`. I can filter these out in the test.

File: tests/shared/core/test_domain.py
Comment (ancalita) [reply tests/shared/core/test_domain.py:1482]: I'll add them 👍

File: rasa/shared/core/generator.py
Comment (ancalita) [reply rasa/shared/core/generator.py]: Yup, removed this part and fingerprinted `self.events` instead.

File: rasa/shared/core/slots.py
Comment (ancalita) [reply rasa/shared/core/slots.py]: I've changed the fingerprinting by using `persistence_info` which should contain the extra bits from certain subclassess (such as `self.values` for `CategoricalSlot`)

File: rasa/shared/core/generator.py
Comment (joejuzl) [inline rasa/shared/core/generator.py:229]: We can just add the whole `domain` as it's fingerprint-able right?

File: rasa/shared/core/generator.py
Comment (joejuzl) [reply rasa/shared/core/generator.py:220]: Well now we don't need the `_states_for_hashing` can we just move this whole implementation to the superclass?

File: tests/shared/core/test_domain.py
Comment (joejuzl) [inline tests/shared/core/test_domain.py:1474]:
```suggestion
    domain_yaml = """
```

File: tests/shared/core/test_events.py
Comment (joejuzl) [inline tests/shared/core/test_events.py:765]: Can we just use `tested_events` here again?

**Review by joejuzl — CHANGES_REQUESTED** [review]

File: rasa/shared/core/generator.py
Comment (ancalita) [reply rasa/shared/core/generator.py:220]: `DialogueStateTracker` doesn't have `self.domain` as attribute. Let me know how best to proceed?

File: rasa/shared/core/generator.py
Comment (ancalita) [reply rasa/shared/core/generator.py:229]: Would adding to the `data` dict this key-value pair work: `{'domain': self.domain.fingerprint()}`?

Comment (joejuzl) [reply rasa/shared/core/generator.py:229]: You can just do `{'domain': self.domain}`. As `deep_container_fingerprint` can now handle `.fingerprint`

File: tests/shared/core/test_events.py
Comment (ancalita) [reply tests/shared/core/test_events.py:765]: Ah yes, good idea!

File: rasa/shared/core/generator.py
Comment (joejuzl) [reply rasa/shared/core/generator.py:220]: True, good catch. 
So the normal way to handle this would be to create `DialogueStateTracker.fingerprint` that fingerprints everything apart from the domain, and then `TrackerWithCachedStates.fingerprint` would add/combine the domain to `super().fingerprint()`. e.g.

```
get_dictionary_fingerprint({"tracker": super().fingerprint(), "domain": self.domain})
```

However, I we may not even need the domain, as I _think_ it's just be used for creating the states. In which case you could just implement it on `DialogueStateTracker.fingerprint`.
@wochinge wdyt?

Comment (ancalita) [reply rasa/shared/core/generator.py:220]: Yikes, it seems `mypy` doesn't allow adding an object to the dictionary, I got this CI error:
```
Dict entry 0 has incompatible type "str": "Domain"; expected "str": "str"  [dict-item]
```
I'll wait for the final decision on whether to move this method to the parent class before making any changes.

File: rasa/shared/core/events.py
Comment (wochinge) [inline rasa/shared/core/events.py:321]: does it make sense to replace / combine this with the existing `__hash__` implementation which we have on every event anyway?

File: rasa/shared/core/generator.py
Comment (wochinge) [reply rasa/shared/core/generator.py:220]: > However, I we may not even need the domain, as I think it's just be used for creating the states. In which case you could just implement it on DialogueStateTracker.fingerprint.

I agree. As soon as the parent node which provides the domain has a changed output this will lead to a "fingerprint miss" anyway

Comment (ancalita) [reply rasa/shared/core/generator.py:220]: Ok, will remove the domain and move the implementation to the parent class.

File: rasa/shared/core/events.py
Comment (ancalita) [reply rasa/shared/core/events.py:321]: Hmm, I steered away from anything using python `hash()` because it's unstable, hence I don't think it's a good idea to use this.

Comment (wochinge) [reply rasa/shared/core/events.py:321]: yep, but our custom `__hash__` implementations in the events, are stable, aren't they? (they should probably not be called `__hash__` in the first place

Comment (ancalita) [reply rasa/shared/core/events.py:321]: Not sure, looking at [`UserUttered`](https://github.com/RasaHQ/rasa/blob/071d2ef24760f738590971c044c703cc27efb567/rasa/shared/core/events.py#L474) and `SlotSet` events for example, their custom `__hash__` uses the built-in `hash()` 🤔  While a few other events like `AllSlotsReset` have [static hashes](https://github.com/RasaHQ/rasa/blob/071d2ef24760f738590971c044c703cc27efb567/rasa/shared/core/events.py#L1046).

I'm also confused because the parent class doesn't have a `__hash__` method.

Comment (wochinge) [reply rasa/shared/core/events.py:321]: What a mess 🙈 The neat thing is that the current `__hash__` implementation excludes all properties which aren't relevant which makes me think of one edge case: Don't get `Event`s assigned timestamps when being read? And wouldn't they differ depending on when you read the training data with the events?

Comment (ancalita) [reply rasa/shared/core/events.py:321]: > Don't get Events assigned timestamps when being read? And wouldn't they differ depending on when you read the training data with the events?

Yes, if timestamp is not specified, then it will get the current timestamp. Although not sure I understand the 2nd question, when would the events be re-instantiated? I am probably not seeing the full implications of where events get used.

However if timestamp is a problem, I can remove it from the fingerprint in the interface.
I still wouldn't use the __hash__ implementation, but if yu think it's a must, I can also spend some effort to create custom fingerprint methods for certain events?

Comment (ancalita) [reply rasa/shared/core/events.py:321]: Yup I just tested using `__hash__` in the fingerprint and it fails the consistency test.

Comment (joejuzl) [reply rasa/shared/core/events.py:321]: We could potentially unify `__hash__` and `fingerprint`. I.e. make `__hash__` use your new stable implementation.

File: tests/shared/core/test_events.py
Comment (joejuzl) [inline tests/shared/core/test_events.py:751]: What exactly is this testing? I think that the problem with inbuild `hash` is only apparent when starting a new python process.

File: tests/shared/core/test_events.py
Comment (joejuzl) [inline tests/shared/core/test_events.py:768]: Tobias makes a good point about `timestamp`. Does this mean that when we re-read the same training data file the event fingerprints won't match?

File: rasa/shared/core/trackers.py
Comment (joejuzl) [inline rasa/shared/core/trackers.py:875]: Event also has `fingerprint` method so can just add it directly.

**Review by joejuzl — CHANGES_REQUESTED** [review]

File: rasa/shared/core/events.py
Comment (ancalita) [reply rasa/shared/core/events.py:321]: By that do you mean delete current `__hash__` body in each individual subclass and return instead `self.fingerprint()`?
Or maybe make `__hash__` use `deep_container_fingerprint()` instead of builtin `hash`?

File: tests/shared/core/test_events.py
Comment (ancalita) [reply tests/shared/core/test_events.py:751]: I thought it's testing if an identical event (i.e, with the same data) returns the same fingerprint 😅 
I just tested with `copy.deepcopy` instead and the test passes.

File: tests/shared/core/test_events.py
Comment (ancalita) [reply tests/shared/core/test_events.py:768]: I don't know enough of what reading the training data file entails to be able to answer, but if timestamp is different every time the training data file is re-read, I will remove the timestamp from the fingerprint.

File: rasa/shared/core/events.py
Comment (ancalita) [inline rasa/shared/core/events.py:1758]: Had to implement separate fingerprint method because the `as_dict` methods for `LegacyForm` and `LegacyFormValidation` update their `type_name` to be identical to their parent class `type_name` and therefore could run into a problem where 2 different event types have the same fingerprint.

File: rasa/shared/core/events.py
Comment (joejuzl) [reply rasa/shared/core/events.py:321]: > By that do you mean delete current __hash__ body in each individual subclass and return instead self.fingerprint()?

Yup exactly. They are both attempting to the do the same thing, i.e. a unique representation.

For reference:
The `__hash__` method is used when you pass the object into the `hash` function and also, in turn, for things like using the object as a key in a `dictionary`, or putting it in a `set`.

File: rasa/shared/core/events.py
Comment (joejuzl) [reply rasa/shared/core/events.py:1758]: Can you put a comment explaining this?

File: rasa/shared/core/events.py
Comment (joejuzl) [inline rasa/shared/core/events.py:1760]: Maybe clearer to re-set the name here rather than calling `super()`?
It took me a second to figure out why this fixed it.

**Comment by joejuzl** [comment]
> I thought it's testing if an identical event (i.e, with the same data) returns the same fingerprint 😅
I just tested with copy.deepcopy instead and the test passed

(For some reason I can't reply to this directly)

Yup I just meant that this won't test the instability of the builtin `hash`. As that only changes when starting a new python _process_.

**Review by joejuzl — CHANGES_REQUESTED** [review]

File: rasa/shared/core/events.py
Comment (wochinge) [reply rasa/shared/core/events.py:321]: > though not sure I understand the 2nd question, when would the events be re-instantiated? I am probably not seeing the full implications of where events get used.

Sorry, I'm a bit short on time, so I hope phrasing this as question works 🙈 
If you read a story file and create a tracker from the events and then do the same thing a couple of seconds later:
Is the fingerprint of both trackers the same?

Comment (ancalita) [reply rasa/shared/core/events.py:321]: @wochinge I'll write a test for that 😄

Comment (ancalita) [reply rasa/shared/core/events.py:321]: @wochinge Yup, timestamp does affect the fingerprint of both the trackers. I'll keep timestamp out of the event fingerprint. And also push this new test too.

Comment (ancalita) [reply rasa/shared/core/events.py:321]: @joejuzl regarding the __hash__ implementation: would it be ok to add __hash__ to parent class and then remove the __hash__ for each individual subclass? There would be too many duplicate return lines if I amend this in every subclass 😄

Comment (ancalita) [reply rasa/shared/core/events.py:321]: ah in the end seems I have to go down this route of duplication otherwise I get a `TypeError: unhashable type` when trying to use hash() on specific event subclasses.

Comment (joejuzl) [reply rasa/shared/core/events.py:321]: > @joejuzl regarding the __hash__ implementation: would it be ok to add __hash__ to parent class and then remove the __hash__ for each individual subclass? There would be too many duplicate return lines if I amend this in every subclass 😄

Yes definitely.

> I get a TypeError: unhashable type when trying to use hash() on specific event subclasses.

huh that's weird. But `fingerprint` returns a `str` right?  which should also always be hashable. Worth looking into a but further I think.

Comment (ancalita) [reply rasa/shared/core/events.py:321]: I think I found the answer [here](https://docs.python.org/3/reference/datamodel.html#object.__hash__) which is identical to Events scenario:
```
If a class that overrides __eq__() needs to retain the implementation of __hash__() from a parent class, the interpreter must be told this explicitly by setting __hash__ = <ParentClass>.__hash__.
```
So either way I still have a lot of lines to duplicate - which one do you think makes most sense? The above one or having a __hash__ implementation per subclass?

File: rasa/shared/core/events.py
Comment (ancalita) [inline rasa/shared/core/events.py:1542]: Had to override `Event.fingerprint()` because a [story conflict test ](https://github.com/RasaHQ/rasa/blob/3f3c8830a99278578ea687bdfc3645d7c368e86b/tests/core/training/test_story_conflict.py#L139) relied on a constant hash for ACTION_UNLIKELY_INTENT ([here ](https://github.com/RasaHQ/rasa/blob/3f3c8830a99278578ea687bdfc3645d7c368e86b/rasa/core/training/story_conflict.py#L248)and [here](https://github.com/RasaHQ/rasa/blob/3f3c8830a99278578ea687bdfc3645d7c368e86b/rasa/core/training/story_conflict.py#L254)).

File: rasa/shared/core/trackers.py
Comment (joejuzl) [inline rasa/shared/core/trackers.py:874]:
```suggestion
            data["events"] = list(self.events)
```

**Review by joejuzl — APPROVED** [review]
Nice! 🚀

**Pull request merged**
